controllers/storagecluster/resourceprofile.go

```diff
@@ -0,0 +1,133 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	resourceProfileChangeApplyingMessage = "New resource profile is being applied"
+	resourceProfileChangeFailedMessage   = "New resource profile failed to apply, please revert to the last working resource profile"
+	resourceProfileChangeTimeout         = 20 * time.Minute
+)
+
+var (
+	resourceProfileChangeStartTime   time.Time
+	errResourceProfileChangeApplying = errors.New(resourceProfileChangeApplyingMessage)
+	errResourceProfileChangeFailed   = errors.New(resourceProfileChangeFailedMessage)
+)
+
+func (r *StorageClusterReconciler) ensureResourceProfileChangeApplied(sc *ocsv1.StorageCluster) error {
+	currentResourceProfile := sc.Spec.ResourceProfile
+	lastResourceProfile := sc.Status.ResourceProfile
+
+	if resourceProfileChangeStartTime.IsZero() {
+		resourceProfileChangeStartTime = time.Now()
+	}
+	// if the resource profile change is taking more than 20 minutes, then say it failed
+	if time.Since(resourceProfileChangeStartTime).Minutes() > resourceProfileChangeTimeout.Minutes() {
+		r.Log.Error(errResourceProfileChangeFailed, "current", currentResourceProfile, "last", lastResourceProfile)
+		return errResourceProfileChangeFailed
+	}
+
+	r.Log.Info("Applying new resource profile", "current", currentResourceProfile, "last", lastResourceProfile)
+
+	// Verify if expected number of mgr pods with the current resource profile label are running
+	if err := r.verifyDaemonWithResourceProfile(sc.Namespace, map[string]string{"app": "rook-ceph-mgr", defaults.ODFResourceProfileKey: currentResourceProfile}, getMgrCount(sc)); err != nil {
+		return err
+	}
+
+	// Verify if expected number of mon pods with the current resource profile label are running
+	if err := r.verifyDaemonWithResourceProfile(sc.Namespace, map[string]string{"app": "rook-ceph-mon", defaults.ODFResourceProfileKey: currentResourceProfile}, getMonCount(sc)); err != nil {
+		return err
+	}
+
+	// Get all the osd pods with the current resource profile label
+	allOsdPodList, err := r.getRunningPodsWithLabels(sc.Namespace, map[string]string{"app": "rook-ceph-osd", defaults.ODFResourceProfileKey: currentResourceProfile})
+	if err != nil {
+		return err
+	}
+	for _, storageDeviceSet := range sc.Spec.StorageDeviceSets {
+		// Get the osds for a particular storage device set
+		osdPodList := &corev1.PodList{}
+		// This regex will find the osd pods with the deviceSet label value as <storageDeviceSet.Name>-<number>
+		pattern := regexp.MustCompile("^" + storageDeviceSet.Name + "-\\d+$")
+		for _, osdPod := range allOsdPodList.Items {
+			if pattern.MatchString(osdPod.Labels["ceph.rook.io/DeviceSet"]) {
+				osdPodList.Items = append(osdPodList.Items, osdPod)
+			}
+		}
+		// Verify if expected number of osd pods with the current resource profile label are running for a particular storage device set
+		count, replica := countAndReplicaOf(&storageDeviceSet)
```

Comment:
Instead of verifying the OSD pod count with the label, how about just checking if there are any pods that don't have the expected resource profile label? Or a similar question, is why are the OSD pods different from checking the mon or mgr pods?

Reply:
Hm, Initially I thought about verifying them each storage device set wise, but seems like it's not really required & just adding complexity.